storage/couchdb/couchdb.go

```diff
@@ -0,0 +1,157 @@
+package couchdb
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/docker/go-connections/tlsconfig"
+	"github.com/flimzy/kivik"
+	couch "github.com/go-kivik/couchdb"
+	"github.com/go-kivik/couchdb/chttp"
+	"github.com/sirupsen/logrus"
+)
+
+// ID of a document in CouchDB; includes Rev for updating an existing
+// document
+type ID struct {
+	ID  string `json:"_id,omitempty"`
+	Rev string `json:"_rev,omitempty"`
+}
+
+// Timing can be embedded into other modesl to add time tracking fields
+type Timing struct {
+	CreatedAt time.Time `json:"created_at"`
+	UpdatedAt time.Time `json:"updated_at"`
+	DeletedAt time.Time `json:"deleted_at"`
+}
+
+func createConnection(tlsOpts tlsconfig.Options, host string, username, password string) (*kivik.Client, error) {
+	// remove username and password from URL to avoid authentication step
+	// before we can set the TLS transport
+	_host, err := url.Parse(host)
+	if err != nil {
+		return nil, err
+	}
+	_host.User = nil
+
+	logrus.Debugf("attempting to connect %s to host %s", username, _host.String())
+
+	client, err := kivik.New(context.TODO(), "couch", _host.String())
+	if err != nil {
+		return nil, err
+	}
+
+	t, err := tlsconfig.Client(tlsOpts)
+	if err != nil {
+		return nil, err
+	}
+	// We have to use the original CipherSuites to be able to create a
+	// connection with CouchDB
+	t.CipherSuites = tls.Config{}.CipherSuites
+
+	setXport := couch.SetTransport(&http.Transport{
+		TLSClientConfig: t,
+	})
+
+	if err = client.Authenticate(context.TODO(), setXport); err != nil {
+		return nil, err
+	}
+
+	if username != "" {
+		var basicAuth chttp.Authenticator = &chttp.BasicAuth{
+			Username: username,
+			Password: password,
+		}
+		if err = client.Authenticate(context.TODO(), basicAuth); err != nil {
+			return nil, err
+		}
+	}
+	return client, nil
+}
+
+// AdminConnection sets up an admin CouchDB connection to the host (`host:port` format)
```

Comment:
I shamelessly recycled that from RethinkDB's usage in cmd/notary-server/config.go:

```
                if doBootstrap {
                        sess, err = rethinkdb.AdminConnection(tlsOpts, storeConfig.Source)
                } else {
                        sess, err = rethinkdb.UserConnection(tlsOpts, storeConfig.Source, storeConfig.Username, storeConfig.Password)
                }
```

Reply:
Ah ok, for RethinkDB that is because if no username is supplied the admin user is used, and by default the admin user has no password.  This seems problematic.  I've filed https://github.com/theupdateframework/notary/issues/1366 to fix this.  If both functions need to take a username and password, it'd probably be better just to have a `Connection` function in the `couchdb` package.